cmd/kubectl-gadget/biolatency.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+)
+
+const (
+	GADGET_NAME        = "biolatency"
+	GADGET_OUTPUTMODE  = "Status"
+	GADGET_OUTPUTSTATE = "Completed"
+)
+
+var biolatencyTraceConfig = &utils.TraceConfig{
+	GadgetName:       GADGET_NAME,
+	TraceOutputMode:  GADGET_OUTPUTMODE,
+	TraceOutputState: GADGET_OUTPUTSTATE,
+	CommonFlags:      &params,
+}
+
+var biolatencyCmd = &cobra.Command{
+	Use:   "biolatency",
+	Short: "Generate a histogram with the distribution of block device I/O latency",
+	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+		// Ensure parent's PersistentPreRunE is still called
+		if cmd.Parent() != nil && cmd.Parent().PersistentPreRunE != nil {
+			err := cmd.Parent().PersistentPreRunE(cmd.Parent(), args)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Biolatency does not support filtering so we need to avoid adding
+		// the default namespace configured in the kubeconfig file.
+		if params.Namespace != "" && !params.NamespaceOverridden {
+			params.Namespace = ""
+		}
+
+		return nil
+	},
+}
+
+var biolatencyStartCmd = &cobra.Command{
+	Use:          "start",
+	Short:        "Start monitor the block device I/O (disk I/O) and records the distribution of I/O latency (time)",
+	RunE:         runBiolatencyStart,
+	SilenceUsage: true,
+}
+
+var biolatencyStopCmd = &cobra.Command{
+	Use:          "stop <trace-id>",
+	Short:        "Stop monitoring and generate a report (a histogram graph) with the distribution of block device I/O latency",
+	RunE:         runBiolatencyStop,
+	SilenceUsage: true,
+}
+
+var biolatencyListCmd = &cobra.Command{
+	Use:          "list",
+	Short:        "List the currently running biolatency traces",
+	RunE:         runBiolatencyList,
+	SilenceUsage: true,
+}
+
+func init() {
+	biolatencyCmd.AddCommand(biolatencyStartCmd)
+	biolatencyCmd.AddCommand(biolatencyStopCmd)
+	biolatencyCmd.AddCommand(biolatencyListCmd)
+
+	utils.AddCommonFlags(biolatencyCmd, &params)
```

Comment:
I think this should be applied on `biolatencyStartCmd`. Common flags are meaningless for list and stop commands.

Reply:
That's currently true. However, given that we are considering adding the support of JSON output, so upon adding that support, the common flags would be useful for the `stop` command as well. 

Said that, do you suggest adding it to `biolatencyStartCmd` in any case? If so, do you want me to revert [this commit](https://github.com/kinvolk/inspektor-gadget/pull/389/commits/1931a7084610c783edc5a3cc1abdec487d08ced3)?

Reply:
> That's currently true. However, given that we are considering adding the support of JSON output, so upon adding that support, the common flags would be useful for the `stop` command as well.

I think we'll need to go further in the detail there. I don't want to have flags in a command that are not relevant. I know the `--output` and related ones are useful for the stop command but all the other ones (that define the filter) aren't. 

> Said that, do you suggest adding it to `biolatencyStartCmd` in any case? 

Yes 

> If so, do you want me to revert [this commit](https://github.com/kinvolk/inspektor-gadget/pull/389/commits/1931a7084610c783edc5a3cc1abdec487d08ced3)?

Yes please!

Reply:
Done

Reply:
I just don't like the fact that right now the `stop` and `list` commands will still accept the `--namespace` flag but they will just ignore it.

Reply:
Yes, we have this problem everywhere as the namespace flag is coming from `KubernetesConfigFlags`. For some reason I think the namespace is misplaced... For instance:

```
$ ./kubectl-gadget execsnoop --help
Trace new processes

Usage:
  kubectl-gadget execsnoop [flags]

Flags:
  -A, --all-namespaces         Show data from pods in all namespaces
  -c, --containername string   Show only data from containers with that name
  -h, --help                   help for execsnoop
      --node string            Show only data from pods running in that node
  -o, --output string          Output format (columns, json, custom-columns). (default "columns")
  -p, --podname string         Show only data from pods with that name
  -l, --selector string        Labels selector to filter on. Only '=' is supported (e.g. key1=value1,key2=value2).
      --verbose                Print additional information

Global Flags:
      --as string                      Username to impersonate for the operation
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string               Default cache directory (default "/home/mvb/.kube/cache")
      --certificate-authority string   Path to a cert file for the certificate authority
      --client-certificate string      Path to a client certificate file for TLS
      --client-key string              Path to a client key file for TLS
      --cluster string                 The name of the kubeconfig cluster to use
      --context string                 The name of the kubeconfig context to use
      --insecure-skip-tls-verify       If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string              Path to the kubeconfig file to use for CLI requests.
  -n, --namespace string               If present, the namespace scope for this CLI request
      --request-timeout string         The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                  The address and port of the Kubernetes API server
      --tls-server-name string         Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                   Bearer token for authentication to the API server
      --user string                    The name of the kubeconfig user to use
```

I'd expect to find `--namespace` together with other filter options, but it's at the end with the Global Flags. I have no idea how to solve it.